h2o-algos/src/main/java/hex/StackedEnsembleModel.java

```diff
@@ -119,7 +114,14 @@ public void initMetalearnerParams(MetalearnerAlgorithm algo) {
     public StackedEnsembleOutput(Job job) { _job = job; }
     // The metalearner model (e.g., a GLM that has a coefficient for each of the base_learners).
     public Model _metalearner;
-    public Frame _levelone_frame_id;
+    public Frame _levelone_frame_id; //set only if StackedEnsembleParameters#_keep_levelone_frame=true
+    public StackingStrategy _stacking_strategy;
+    
+    //Set of base model predictions that have been cached in DKV to avoid scoring the same model multiple times,
+    // it is then the responsibility of the client code to delete those frames from DKV.
+    //This especially useful when building SE models incrementally (e.g. in AutoML).
+    //The Set is instantiated and filled only if StackedEnsembleParameters#_keep_base_model_predictions=true.
+    public NonBlockingHashSet<String> _base_model_predictions_keys; 
```

Comment:
Is there a good reason to store Keys as Strings? Why a HashSet? It would be good to be able to link the prediction frame back to the model - using an array and preserving the same order of models would enable it.

Reply:
@michalkurka
Regarding the Hashset, I just wanted to ensure there's no duplicate: predictions are not only associated to a model, but to a model+frame combination; in CV-stacking mode, this frame can only be the validation frame, but in blending mode, it could be the blending frame or the validation frame. 
For now, the purpose of those keys is only to track them for deletion later. The associated predictions are not exposed through REST API (validation frame predictions also were not until now): if we want to expose them, then we can think about a way, but I don't think that simply sorting by model order would be convenient enough anyway. 
From backend, it is currently as simple to do a reverse lookup: given list of models, a prediction key and a frame, build model prediction keys for the model+frame until it matches the given key... not extremely efficient, but as there's no use-case currently anyway...

Regarding the String key: In a early implementation, I started with a Key but I had some issue with serialization then, so I switched to String... will try again, but it may be related with the use of a NonBlockingHashSet... this Iced mechanism is actually seriously limiting our usage of data structures.

Reply:
@michalkurka : now that client mode tests are failing, I know why it used to be String instead of key:
```
Caused by: java.io.InvalidClassException: water.Key; no valid constructor
	at java.io.ObjectStreamClass$ExceptionInfo.newInvalidClassException(ObjectStreamClass.java:169)
	at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:874)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2043)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at water.nbhm.NonBlockingHashMap.readObject(NonBlockingHashMap.java:1421)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1170)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2178)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at water.AutoBuffer.javaSerializeReadPojo(AutoBuffer.java:1514)
```

I believe this is because the NonBlockingHashset is only Serializable... so, going to replace this with an array tmrw...